File: 04/js/part1.js
const fs = require('fs')

const contents = fs.readFileSync('input.txt', 'utf8')

const passphrases = contents.split('\n').filter(line => line !== '')

const isValid = passphrase => {
  const wordsSeenSoFar = new Set()
  const wordsInThisPassphrase = passphrase.split(' ')
  for (const word of wordsInThisPassphrase) {
    if (wordsSeenSoFar.has(word)) {
      return false
    }
    wordsSeenSoFar.add(word)
  }
  return true
}

const valid = passphrases.filter(isValid)

console.log(valid.length)

File: 04/hs/part1.hs
import qualified Data.Set as Set
import Data.List

valid :: String -> Bool
valid "" = True
valid p = Set.size(Set.fromList ws) == length ws
  where ws = words p

main = do
    inputFile <- readFile "../input.txt"
    let passphrases = filter (/= "") $ lines inputFile
    let validPassphrases = filter valid passphrases
    print $ length validPassphrases
